doc/dev/templates.md

```diff
@@ -0,0 +1 @@
//
// Example config:
// 
// apiVersion: "config.istio.io/v1alpha2"
// kind: listentry
// metadata:
//   name: appversion
//   namespace: istio-config-default
// spec:
//   value: source.labels["version"]
message Template {
    // Specifies the entry to verify in the list.
    string value = 1;
}
</pre>

The interesting parts of this definition are:

- **Package Name**. The package name determines the name by which the template will be known, both by
adapter authors and operators using adapters that implement the template. 
The name should be written in camelCase.

- **Variety**. The variety of a template determines at what point in the Mixer's processing pipeline the
adapters that implement the template will be called. This can be CHECK, REPORT, or QUOTA. 

- **Message Name**. The name of the template message should always be `Template`

- **Fields**. The fields represent the general shape of the data that will be delivered to the adapters at
runtime. The operator will need to populate these fields via configuration.

## Supported field names

Template fields can have any valid protobuf field name, except for the reserved name `name`.

## Supported field types

Templates currently only support a subset of the full protobuf type system. Fields in a template can
be one of the following types:

- `string`
- `int64`
- `double`
- `bool`
- `istio.mixer.v1.config.descriptor.ValueType`
- `google.protobuf.Timestamp`
- `google.protobuf.Duration`
- `map<string, string>`
- `map<string, int64>`
- `map<string, double>`
- `map<string, bool>`
- `map<string, google.protobuf.Timestamp>`
- `map<string, google.protobuf.Duration>`
- `map<string, istio.mixer.v1.config.descriptor.ValueType>`

There is currently no support for nested messages, enums, `oneof`, and `repeated`.

The type `istio.mixer.v1.config.descriptor.ValueType` has a special meaning. Use of this type
tells Mixer that the associated value can be any of the supported attribute
types are defined by the [ValueType](https://github.com/istio/api/blob/master/mixer/v1/config/descriptor/value_type.proto)
enum. The specific type that will be used at runtime depends on the configuration the operator writes.
Adapters are told what these types are at configuration time so they can prepare themselves accordingly.

## Adding a template to Mixer

Templates are statically linked into Mixer. To add or modify a template, it is therefore necessary to produce a new
Mixer binary.

** TBD **

## Template evolution compatibility

Here's what can be done to a template and remain backward compatible with existing adapters and operator configurations that use the
template:

- Adding a field

The following changes cannot generally be made while maintaining compatibility:

- Renaming the template
- Changing the template's variety
- Removing a field
- Renaming a field
- Changing the type of a field

```

Comment:
why so many "^M"s here?

Reply:
Cannot select a line to comment so ...
Mixer determines... paragraph
---
Mixer calls adapters with request arguments during request processing. The specific structure of request args is parameterized as a `template`.  
A template serves two purposes
1. Defines the structure of the specific request arg that is sent to the adapter.
2. Defines how an operator configures the mapping from attributes to the structure defined in 
   `#1`
--